### ISSUE TITLE
Sporadically app doesn't load on fresh install (iOS)

### ISSUE THREAD

**Issue by rajatrocks** [markmarijnissen/cordova-app-loader#17]
When I remove my app from my device and reinstall/run it via Xcode, about one time in 7 it won't load and gets stuck on the splash screen. The last thing Xcode shows in its log is: 

2015-01-28 15:59:56.290 MyApp[1960:506893] Finished load of: file:///private/var/mobile/Containers/Bundle/Application/F541E2A0-4247-4ACF-B80D-2603890E70A1/MyApp.app/www/index.html

When the app runs properly, after that I see several log messages from my app. 

Looking at the Safari debugger, I can see that all the script and css nodes were properly added by bootstrap.js, it just appears that none of the code is being run - it's unclear why.

- Removing Cordova-app-loader fixes the problem (although that's clearly not desirable). 
- Adding alert ("IN BOOTSTRAP.JS") to the top of bootstrap.js also fixes the problem. Because of that and the sporadic nature of the issue, it would seem to point to some sort of timing problem that the time taken to close the alert addresses. 

It only ever seems to happen on the first run after a fresh install. I can't get it to happen on any subsequent runs. 

I am using the 0.14.0 version of the library. 

I have tested and can replicate this on an iPad 1, an iPad Mini 2, and various iOS simulator devices. I don't have this problem on Android. 

Please let me know if I can provide you with any other information, or if you have suggestions on where in the code I can look to try to narrow down the issue. 

Thanks!

**Comment by lylepratt** [comment]
I have not personally experienced this, but other members of my team have. They have all been using iOS.

**Comment by rajatrocks** [comment]
btw - the load part of my manifest.json looks like this: 

  "load": [
    "bootstrap/cordova-app-loader-complete.min.js",
    "styles/vendor.css",
    "styles/main.css",
    "scripts/vendor.js",
    "scripts/scripts.js"
  ],

**Comment by rajatrocks** [comment]
Good to know it's not just me @lylepratt !

**Comment by rajatrocks** [comment]
I wonder if the UIWebView that Cordova iOS apps use doesn't actually support script.async=false (as implemented in https://github.com/markmarijnissen/cordova-app-loader/issues/6), since it's different than iOS Safari? That would explain the problem, since then the app would be doing 
- *I understand “async”, but don’t understand setting it to “false” with JS. I’ll execute your scripts as soon as they land, in whatever order* 
instead of 
- *Download together, execute in order as soon as all download.*.

**Comment by lylepratt** [comment]
I'm fairly certain that is not the problem.

**Comment by rajatrocks** [comment]
@lylepratt @markmarijnissen - Any thoughts on where in the code I should be looking to narrow down the cause of the problem? I had to remove Cordova App Loader from my project for the time being in order to submit my app for review, but would love to get the functionality working.

**Comment by lylepratt** [comment]
I believe its related to the time it takes for the app loader to copy assets over from the App bundle to the local manifest location.

**Comment by rajatrocks** [comment]
Diving into this today, and have some questions: 
@markmarijnissen 
- have you seen this problem before? 
- is your app built using angular?
- am I correct that if you don't put any code in your app to call CordovaAppLoader, that it doesn't do anything on its own? Bootstrap.js loads, loads the manifest, then loads all the "load" files listed (including cordova-app-loader-complete.js). The cordova-app-loader-complete.js file loads all the 6 modules, but none of them do anything except setting some variables, until they are explicitly called by the app. So on first run, none of this does anything special. 

@lylepratt
- is your app built using angular?

**Comment by lylepratt** [comment]
@rajatrocks yes, we use angular.

**Comment by rajatrocks** [comment]
My app is also in Angular. After more tracing and debugging, here's what I believe is happening: 

- If you use the ng-app directive (which I do) then Angular automatically bootstraps the app when the page thinks all the content is loaded: https://docs.angularjs.org/guide/bootstrap
- My app has 2 JS files it loads: vendor.js has angular and other framework code and is loaded first. Then scripts.js has all my app logic. 
- bootstrap.js looks at the manifest and then loads vendor.js (which has angular), which then tries to bootstrap the angular app as soon as it loads. If scripts.js hasn't been loaded by the time it tries to do this, the angular bootstrap fails and the app never loads. That's what's causing the problem and why it only happens sometimes. If scripts.js has been loaded, then everything works fine.
- I'm guessing I only see this on first run after a fresh install because after that both vendor.js and scripts.js might be in the webview cache, so loading happens instantly. 

**Potential Solutions:** 
- Combine all js into one file, so you guarantee that everything is loaded. I tried this as an experiment and the app loaded fine every time. I haven't been able to get it to hang on the splash screen and I've uninstalled and then reinstalled and run it 30+ times. 
- Manually initialize angular after everything has been loaded, which is exactly what they suggest for when using "script loaders". The question then becomes when? Some potential ideas: 
  - In bootstrap.js in the loadScripts() method, add an array that tracks each JS file that is being loaded, and add an onload eventhandler to each script node you add. In the onload handler, mark that JS file as completed, and when all the JS files have been loaded, only then call angular.bootstrap(). So you're guaranteed that all the scripts have loaded before angular starts up.
    - Could make this more generic (for people who do and don't use angular) by implementing the onload scheme above, and then when all the JS files are loaded, if a window.calBootstrapComplete function exists, call it. That way people not using angular don't need to do anything, and those that are can add the calBootstrapComplete function to their index.html and use it to call angular.bootstrap(). 
    - Or you can possibly specify the name of the function or whether to call it in the bootstrap.js script tag, the same way the manifest and timeout are specified today. 
    - This option seems like the best to "productize". 
  - Alternatively if you didn't want to touch the cordova-app-loader code, you could put a function in index.html that looked for the existence of content that you know is in the JS files that are being loaded (on a timer?), and once it finds everything it needs, it calls angular.bootstrap(). Or put in special code in your files that sets variables on the window object that the function in index.html looks for. Or something to that effect. That would be custom for everyone. Seems like a lousy solution, just logging it as an alternative.  
  - Any other ideas?

**Comment by rajatrocks** [comment]
So I ended up implementing the suggested solution above with some minor modifications. This solution supports loading multiple JS files, using the callback or not, and enables you to specify your callback function.

### index.html
Add a "callback" attribute to the bootstrap script with the name of the callback function. I put it here instead of in the manifest because I need to be able to call it when there's no manifest - which is how my development environment is setup. 
`<script type="text/javascript" timeout="10000" manifest="manifest.json" src="bootstrap/bootstrap.js" callback="calBootstrapComplete"></script>`

I had the ng-app directive on the body tag. I removed it and added the callback function that bootstraps angular.
```javascript 
        <script>
            // Startup angular. 
            function calBootstrapComplete() {
                angular.element(document).ready(function() {                    
                    angular.bootstrap(document.body, ['MyApp']);
                    // need this for e2e testing: https://github.com/karma-runner/karma/issues/422
                    document.body.className = document.body.className + ' ng-app';
                });
            };      
        </script>
```
### bootstrap.js
Pull the callback from the script tag and save it in a variable, and immediately call it if there's no manifest. 
```javascript
//---------------------------------------------------------------------
window.Manifest = {};
// Step 1: Load manifest from localStorage
var manifest = JSON.parse(localStorage.getItem('manifest'));
// grab manifest.json location from <script manifest="..."></script>
var s = document.querySelector('script[manifest]');
// ********* NEW LINES
var callback = s.getAttribute('callback'); 
// ********* END NEW LINES

// Not in localStorage? Fetch it!
if(!manifest){
  var url = (s? s.getAttribute('manifest'): null) || 'manifest.json';
  // get manifest.json, then loadManifest.
  pegasus(url).then(loadManifest,function(xhr){
    console.error('Could not download '+url+': '+xhr.status);
    // ********* NEW LINES
    // no manifest, callback immediately. 
    if (callback) {
        window[callback]();
    }
    // ********* END NEW LINES
  });
// Manifest was in localStorage. Load it immediatly.
} else {
  loadManifest(manifest,true,s.getAttribute('timeout') || 10000);
}
})();
```

And modify loadScripts to track each script and call the callback when they've all been loaded. 
```javascript
  function loadScripts(){
    // ********* NEW LINES
    var jsToLoad = [];
    // ********* END NEW LINES
    scripts.forEach(function(src) {
      if(!src) return;
      // Ensure the 'src' has no '/' (it's in the root already)
      if(src[0] === '/') src = src.substr(1);
      src = manifest.root + src ;
      // Load javascript
      if(src.substr(-3) === ".js"){        
        el= document.createElement('script');
        el.type= 'text/javascript';
        el.src= src + '?' + now;
        el.async = false;
        // ********* NEW LINES        
        if (callback) {        
            jsToLoad.push(src);            
            el.onload = function() {                
                var i = jsToLoad.indexOf(src);
                if (i != -1) {                    
                    jsToLoad.splice(i,1);
                }
                if (jsToLoad.length == 0) {                    
                    window[callback]();
                }
                this.onload = null;
            };
        }
        // ********* END NEW LINES  
      // Load CSS
      } else {
        el= document.createElement('link');
        el.rel = "stylesheet";
        el.href = src + '?' + now;
        el.type = "text/css";
      }
      head.appendChild(el);
    });
  }
```

**Comment by lylepratt** [comment]
I've verified that this continues to happen. And it only ever happens the FIRST time an app tries to load after being installed. Can't figure out why though.

**Comment by rajatrocks** [comment]
Interesting. I can't get it to happen anymore after making the changes above.

**Comment by techntrails** [comment]
@lylepratt, can you please confirm if your previous comment relates to @rajatrocks's recent solution? I'd like to try his solution but not if the issue you said 'continues to happen' is happening w/ his fix in place.

**Comment by rajatrocks** [comment]
fyi - I'm no longer having the loading problems, but am currently encountering some other issues with updates not being found and CSS being loaded incorrectly on Android. This stuff is hard to debug on device...

**Comment by Urbiwanus** [comment]
> My app has 2 JS files it loads: vendor.js has angular and other framework code and is loaded first. Then scripts.js has all my app logic including module definition, controllers, etc.

To keep track of the right order of the loaded files i added a simple order array to my GruntFile, to generate a complete list of my generated www folder without defining order for each file manually 

https://gist.github.com/Urbiwanus/1594a66a746e60b7b906

**Comment by markmarijnissen** [comment]
Hey everybody!

Thanks you all for your contribution to this issue. 

@rajatrocks, thanks for your solution. Could you create a Pull Request for this? Then I'll merge it into master. I think it's a valuable addition. 

However, there is also an alternative: As scripts are loaded in order, you can simply put initialization logic into your last script. (In this case, the Angular Bootstrap call)

@Urbiwanus Bootstrap.js will load the scripts in the exact order you specify in the Manifest.

**Comment by rajatrocks** [comment]
Hi @markmarijnissen - I will try to learn how to do a Pull Request in the next few days and then make one with my changes.  

btw - @Urbiwanus has discovered and come up with solutions for some other problems with using CordovaAppLoader with Angular.

**Comment by lylepratt** [comment]
I am now encountering this sporadically even after first load. My solution so far is to bundle all my JavaScript together and I plan to submit an update with that tomorrow. However, its gotten me kinda scared because I don't know why its happening not just on first load now.

**Comment by lylepratt** [comment]
Just updating you guys further. Placing all of the JS into the same file did not solve the issue for me. I've implemented @rajatrocks solution and that does appear to solve it. Thanks @rajatrocks for the detailed and notated changes.

**Comment by lylepratt** [comment]
OK. I implemented @rajatrocks solution, but unfortunately I am still receiving white screen after fresh installs of the app.

**Comment by lylepratt** [comment]
I noticed that the `Step 5` in the Bootstrap file never returns true (never waits for Cordova to load). This is why my app would get stuck on the white screen. I changed step 5 to this, and it seems to solve my problem:

```
  if(manifest.root.substr(0,7) === 'cdvfile' || window.cordova){
    document.addEventListener("deviceready", loadScripts, false);
  } else {
    loadScripts();
  }
``
Essentially, if the Cordova library is included in the project, I'm assuming I need to wait for the `deviceready` event.

**Comment by Urbiwanus** [comment]
@lylepratt  Nice 

Please make a PR

**Comment by Urbiwanus** [comment]
I changed line too 

    if(!!window.cordova){
      document.addEventListener("deviceready", loadScripts, false);
    } else {
      loadScripts();
    }

**Comment by markmarijnissen** [comment]
Added the improved cordova check. Hope this fixes it (version 0.15.0)! I not, please re-open the issue.

**Comment by jasonz1987** [comment]
i met the same issue on ios with my ionic app recently,it does't work when use @rajatrocks code,so what's the reason?
